Content.Server/StationEvents/StationEvent.cs

```diff
@@ -49,37 +90,105 @@ public abstract class StationEvent
+        }
+
+        /// <summary>
+        /// Called when the tick is equal to the StartWhen variable.
+        /// </summary>
+        public abstract void Start();
+
+        /// <summary>
+        /// Called when the tick is qual to the AnnounceWhen variable.
+        /// </summary>
+        /// <param name="fake"></param>
+        public virtual void Announce(bool fake)
+        {
             if (StartAnnouncement != null)
             {
                 var chatManager = IoCManager.Resolve<IChatManager>();
                 chatManager.DispatchStationAnnouncement(StartAnnouncement);
             }
+            if (StartAudio != null)
+            {
+                EntitySystem.Get<AudioSystem>().PlayGlobal(StartAudio, AudioParams.Default.WithVolume(-10f));
+            }
+        }
+
+        /// <summary>
+        /// Called once when the station event ends
+        /// </summary>
+        public virtual void End()
+        {
+            if (EndAnnouncement != null)
+            {
+                var chatManager = IoCManager.Resolve<IChatManager>();
+                chatManager.DispatchStationAnnouncement(EndAnnouncement);
+            }
+            if (EndAudio != null)
+            {
+                EntitySystem.Get<AudioSystem>().PlayGlobal(EndAudio, AudioParams.Default.WithVolume(-10f));
+            }
+            Running = false;
         }
 
         /// <summary>
         /// Called every tick when this event is active
         /// </summary>
         /// <param name="frameTime"></param>
-        public abstract void Update(float frameTime);
+        public virtual void Tick(float frameTime)
+        {
+            return;
+        }
 
         /// <summary>
-        /// Called once when the station event ends
+        /// Called every tick when this event is running. Do not override me!
         /// </summary>
-        public virtual void Shutdown()
+        /// <param name="frameTime"></param>
+        public virtual void Update(float frameTime)
         {
-            if (EndAnnouncement != null)
+            if (!Running)
             {
-                var chatManager = IoCManager.Resolve<IChatManager>();
-                chatManager.DispatchStationAnnouncement(EndAnnouncement);
+                return;
+            }
+
+            if (ActiveFor >= StartWhen && !Started)
+            {
+                Start();
+                Started = true;
+            }
+
+            if (ActiveFor >= AnnounceWhen && !Announced)
+            {
+                Announce(false);
+                Announced = true;
+            }
+
+            if (StartWhen < ActiveFor && ActiveFor < EndWhen)
+            {
+                Tick(frameTime);
             }
+
+            if (ActiveFor >= EndWhen && ActiveFor >= AnnounceWhen && ActiveFor >= StartWhen)
+            {
+                End();
+                Started = false;
+                Announced = false;
+                return;
+            }
+            ActiveFor += frameTime;
```

Comment:
IMO move the timer to the top of these methods, that way you can early out rather than needing to check all the others. E.g. if the event hasn't started you can just do the StartWhen < ActiveFor && !Started then return.

Also I think the announcement should occur as soon as the event starts but the event starts after the start timer?

The variables will also need resetting under the end / shutdown method.

If you add a test for the timers under the existing test you also won't need to check all of the timers have been met. Also the shutdown method is already called by the system so you won't need to call it here, you'll just need to set Running to false and the system will handle it.

Reply:
could i just move this to the `StationEventSystem` Update so i can return the `StationEvent`'s Update() back?

Reply:
Maybe, but one advantage here is you can mark stuff like Started as private so other stuff can't touch it, whereas you'd need to make it public if moved.